tests/element_test_vectors.rs

```diff
@@ -37,69 +37,56 @@ type SkipList = &'static [&'static str];
 trait RoundTrip {
     /// Encodes `elements` to a buffer in the specified Ion `format` and then reads them back into
-    /// a `Vec<OwnedElement>` using the provided reader.
+    /// a `Vec<Element>` using the provided reader.
     fn roundtrip<R: ElementReader>(
         elements: &[Element],
         format: Format,
         reader: R,
     ) -> IonResult<Vec<Element>>;
 }
 
-/// These unit structs implement the [RoundTrip] trait and can be passed throughout the
-/// integration tests to dictate which readers/writers should be used in each test.
-struct NativeElementWriterApi;
-
-impl RoundTrip for NativeElementWriterApi {
-    fn roundtrip<R: ElementReader>(
-        elements: &[Element],
-        format: Format,
-        reader: R,
-    ) -> IonResult<Vec<Element>> {
-        // Unlike the C writer, the Rust writer can write into a growing Vec.
-        // No need for an aggressive preallocation.
-        let mut buffer = Vec::with_capacity(2048);
-        match format {
-            Format::Text(kind) => {
-                let writer = match kind {
-                    TextKind::Compact => TextWriterBuilder::new().build(&mut buffer),
-                    TextKind::Lines => TextWriterBuilder::lines().build(&mut buffer),
-                    TextKind::Pretty => TextWriterBuilder::pretty().build(&mut buffer),
-                }?;
-                let mut writer = NativeElementWriter::new(writer);
-                writer.write_all(elements)?;
-                let _ = writer.finish()?;
-            }
-            Format::Binary => {
-                let binary_writer = BinaryWriterBuilder::new().build(&mut buffer)?;
-                let mut writer = NativeElementWriter::new(binary_writer);
-                writer.write_all(elements)?;
-                let _ = writer.finish()?;
-            }
-        };
-
-        reader.read_all(buffer.as_slice())
-    }
+/// Serializes all of the given [Element]s to a `Vec<u8>` according to the
+/// specified [Format].
+fn serialize(format: Format, elements: &[Element]) -> IonResult<Vec<u8>> {
+    let mut buffer = Vec::with_capacity(2048);
+    match format {
+        Format::Text(kind) => {
+            let writer = match kind {
+                TextKind::Compact => TextWriterBuilder::new().build(&mut buffer),
+                TextKind::Lines => TextWriterBuilder::lines().build(&mut buffer),
+                TextKind::Pretty => TextWriterBuilder::pretty().build(&mut buffer),
+            }?;
+            let mut writer = NativeElementWriter::new(writer);
+            writer.write_all(elements)?;
+            let _ = writer.finish()?;
+        }
+        Format::Binary => {
+            let binary_writer = BinaryWriterBuilder::new().build(&mut buffer)?;
+            let mut writer = NativeElementWriter::new(binary_writer);
+            writer.write_all(elements)?;
+            let _ = writer.finish()?;
+        }
+    };
+    Ok(buffer)
 }
 
 trait ElementApi {
-    type ReaderApi: ElementReader;
```

Comment:
🗺️ I was also able to remove this associated type. Now different readers can be configured and returned as the same (dynamically dispatched) `Reader` type.

Reply:
🥳